3단계 - 질문 삭제하기 리펙터링

haneepark opened the pull request:
민성님, JPA 미션의 마지막 단계 리뷰 요청드립니다!

haneepark committed:
refactor(QnaService, Question): 질문 삭제 권한 확인, 삭제 flag 세팅하는 로직을 도메인 엔티티 안으로 이동

haneepark committed:
refactor(QnaService, DeleteHistory): DeleteHistory 생성 로직을 DeleteHistory의 팩토리 메소드 안으로 이동

minsungryu commented on src/main/java/qna/domain/DeleteHistory.java:
```diff
@@ -32,6 +35,45 @@ public DeleteHistory(ContentType contentType, Long contentId, User deletedBy, Lo
+    }
+
+    static DeleteHistory ofQuestion(Question question) {
+        return new DeleteHistory(ContentType.QUESTION, question.getId(), question.getWriter(), LocalDateTime.now());
```
`@CreatedDate` 애너테이션을 활용하면 `createDate`를 인자로 넘겨주지 않아도 되지 않을까요?

haneepark replied:
e32ce4d 에 반영했습니다!

minsungryu commented on src/main/java/qna/domain/Answer.java:
```diff
@@ -55,6 +56,7 @@ public boolean isOwner(User writer) {
 
     public void toQuestion(Question question) {
         this.question = question;
+        question.getAnswers().add(this);
```
양방양 매핑을 해주신 경우 기존에 `question`이 할당되어 있었는지도 검사하는 것은 어떨까요?
참조를 변경하기 전 `answers`에서 제거해주는 작업도 필요해보이는데요
연관 관계의 주인이 `Answer`라서 데이터베이스 상으로는 문제가 없을 수 있지만 엔티티의 객체 그래프는 그렇지 않을 수 있어서 말씀드렸습니다.

haneepark replied:
e32ce4d 에 반영했습니다!

minsungryu commented on src/main/java/qna/domain/Answer.java:
```diff
@@ -107,4 +109,14 @@ public String toString() {
                 '}';
     }
+
+    public void validateDeletableBy(User user) throws CannotDeleteException {
```
- 이 메서드는 `delete()`메서드 내부에서 호출되고 외부에는 공개되면 안되지 않을까요?
- `CannotDeleteException`를 비검사 예외로 만들어보는 것은 어떨까요?

haneepark replied:
> CannotDeleteException를 비검사 예외로 만들어보는 것은 어떨까요?

`RuntimeException`을 상속하도록 구현하라는 말씀이신거죠? 왜 제안하셨는지 설명 해주실수있을까요?

haneepark replied:
> 이 메서드는 delete()메서드 내부에서 호출되고 외부에는 공개되면 안되지 않을까요?

e32ce4d 에 반영했습니다!

minsungryu replied:
클린 코드에서 나온 내용과 JPA 사용시 주의사항이 있어서 말씀드렸는데요,
관련해서는 잘 정리된 포스트로 답변을 대신하도록 하겠습니다 :)

- https://velog.io/@kwj1270/%ED%81%B4%EB%A6%B0-%EC%BD%94%EB%93%9C-%EC%98%A4%EB%A5%98-%EC%B2%98%EB%A6%AC#%EB%AF%B8%ED%99%95%EC%9D%B8unchecked-%EC%98%88%EC%99%B8%EB%A5%BC-%EC%82%AC%EC%9A%A9%ED%95%98%EB%9D%BC
- https://cheese10yun.github.io/checked-exception/

minsungryu commented on src/main/java/qna/domain/DeleteHistory.java:
```diff
@@ -32,6 +35,45 @@ public DeleteHistory(ContentType contentType, Long contentId, User deletedBy, Lo
+    }
+
+    static DeleteHistory ofQuestion(Question question) {
+        return new DeleteHistory(ContentType.QUESTION, question.getId(), question.getWriter(), LocalDateTime.now());
```
또한 메서드 내부에서 `LocalDateTime.now()`를 호출하면 두 가지 문제(?)가 예상되는데요,
1. 시간과 관련된 테스트가 어려울 수 있습니다.
2. `question`과 `answer`의 삭제 시간에 미세한 차이가 있을 수 있습니다.

haneepark replied:
그렇네요 의견 감사합니다
e32ce4d 에 반영했습니다!

minsungryu commented on src/main/java/qna/domain/Question.java:
```diff
@@ -20,6 +24,9 @@
     private boolean deleted = false;
 
+    @OneToMany(mappedBy = "question")
+    private List<Answer> answers = new ArrayList<>();
```
답변을 추가하는 경우에도 양방향 매핑을 고려해서 메서드가 수정되어야 하지 않을까요?
53라인에 연관 관계를 맺어줄 수 있도록 변경이 필요해보입니다.

haneepark replied:
엇 저는 `answer.toQuestion()`을 호출하니까 이쪽에서 다 처리가 되어서 문제없다고 생각했는데요. `toQuestion()` 호출 외에 추가로 처리해주어야 할 것이 있나요?

```java
public void addAnswer(Answer answer) {
    answer.toQuestion(this);
}
```

minsungryu replied:
`Question`이 가진 `answers` 컬렉션에도 추가가 되어야하지 않을까요?

```java
public void addAnswer(Answer answer) {
    answers.add(answer);
    answer.toQuestion(this);
}
```

위 코드 역시 몇 가지 문제점이 있는데요, 양방향 연관 관계 편의 메서드에 대해서 알아보시면 도움이 될 것 같습니다.

minsungryu commented on src/main/java/qna/domain/Question.java:
```diff
@@ -20,6 +24,9 @@
 
     private boolean deleted = false;
 
+    @OneToMany(mappedBy = "question")
```
Cascade, Fetch 전략에 대해서도 고민해보시면 좋을 것 같습니다.

haneepark replied:
Fetch/Cascade 전략 고려해서 커밋 0fefb05 에 수정사항 적용했습니다!

Fetch 전략
* `Question`, `Answer`, `DeleteHistory`에서 각각의 User 필드는 eager fetch. 비즈니스 요구사항(없긴하지만 상상해본다면)으로인해 자주 함께 사용될 것 같아서.
* `Question`의 answers와 `Answer`의 question은 lazy fetch. 함께 덜 사용될 것 같아서.

Cascade 전략
* `Question`의 answers에 `CascadeType.ALL` 설정 추가

minsungryu replied:
하늬님 이유까지 설명해주시고 너무 좋은 것 같습니다 👍 
하나 더 말씀드리면 가능한 `LAZY` 전략을 사용하면서 성능상 문제되는 부분에 대해서는 전략을 수정하거나 다른 문제 해결책을 찾기도 하니 참고하시면 도움이 될 것 같습니다 :)

minsungryu commented on src/main/java/qna/domain/User.java:
```diff
@@ -132,4 +132,17 @@ public boolean isGuestUser() {
+        if (this == o) return true;
+        if (o == null || getClass() != o.getClass()) return false;
+        User user = (User) o;
+        return Objects.equals(id, user.id) && Objects.equals(userId, user.userId);
```
엔티티의 경우는 `id`에 대해서만 동등성을 검사하면 되지 않을까요?
아래 포스트를 읽어보시면 도움이 될 것 같습니다.
https://jaeyeolshin.github.io/2016-02-06/difference-between-entity-and-value-object/

haneepark replied:
e32ce4d 에 반영했습니다!

minsungryu commented on src/test/java/qna/domain/DeleteHistoryTest.java:
```diff
@@ -0,0 +1,40 @@
+
+    @Test
+    void ofAnswer() {
+        final DeleteHistory deleteHistory = DeleteHistory.ofAnswer(AnswerTest.A1);
```
메서드를 테스트하기 위해 가시성을 열어둔 것으로 이해했는데 맞을까요?
비공개 메서드를 테스트 해야하는지에 대해서는 아래 포스트를 읽어보시면 좋을 것 같습니다!
https://justhackem.wordpress.com/2017/09/29/should-private-methods-be-tested/

haneepark replied:
b7b72b9에 ofQuestion, ofAnswer 메서드를 private으로 변경하고 테스트는 제거했습니다.
생각해보니 DeleteHistoryTest의 of() 로도 충분히 검증이 되는데, 불필요한 테스트였던것같네요 :)

minsungryu commented on src/test/java/qna/domain/QuestionTest.java:
```diff
@@ -48,11 +56,54 @@ void findByDeletedFalse() {
+    }
+
+    @Test
+    void validateDeletableBy_댓글_작성자가아닌경우() {
```
테스트명으로 설명하기 어렵지 않으신가요?ㅎㅎ
`@DisplayName` 애너테이션을 활용해보셔도 좋을 것 같습니다.
또한 테스트를 설명할 때 기대하는 바도 있으면 이해하기 더 좋지 않을까 싶습니다.
예시) "댓글 작성자가 아닌 경우 질문을 삭제할 수 없다"

haneepark replied:
넵 DisplayName까지 미처 신경을 못썼네요.. 이번 미션에서는 패스하도록 하겠습니다🥲

minsungryu commented on src/test/java/qna/domain/QuestionTest.java:
```diff
@@ -48,11 +56,54 @@ void findByDeletedFalse() {
+        final User user = users.save(UserTest.JAVAJIGI);
+        final Question question = questions.save(QuestionTest.Q1.writeBy(user));
+
+        question.validateDeletableBy(user);
```
앞서 말씀드린대로 공개 메서드가 아닌 것에 대한 테스트가 필요한지, 필요하다면 어떤 방식으로 테스트해야할 지 한 번 더 고민해보셔도 좋을 것 같습니다.

minsungryu commented on src/test/java/qna/domain/QuestionTest.java:
```diff
@@ -48,11 +56,54 @@ void findByDeletedFalse() {
+        final User user = users.save(UserTest.JAVAJIGI);
+        final Question question = questions.save(QuestionTest.Q1.writeBy(user));
+        answers.save(new Answer(user, question, "Answers Contents1"));
+        entityManager.clear();
```
영속성 컨텍스트를 초기화해서 아래 조회 수행시에 DB에서 값을 불러오도록 의도하신걸까요?

haneepark replied:
맞습니다!!

minsungryu commented on src/test/java/qna/domain/QuestionTest.java:
```diff
@@ -48,11 +56,54 @@ void findByDeletedFalse() {
     }
 
+    @Test
+    void findByDeletedFalse_WithAnswer() { // TODO: 2021/11/25 setAnswer가 되어야지!!!
```
요 TODO는 무엇일까요 ㅎㅎ

haneepark replied:
사실.. 기억이..안납니다... 실수로 남겨뒀네요 삭제하겠습니다😅

minsungryu reviewed (changes requested):
하늬님, 3단계 구현 잘 해주셨네요!
코드가 깔끔하고 이해하기 쉬웠습니다 👍 
양방향 매핑과 테스트 관련해서 몇 가지 코멘트를 남겼는데요
확인해주시고 다시 리뷰 요청주시면 좋을 것 같습니다 :)

haneepark committed:
refactor: 코드 리뷰 의견 반영

- DeleteHistory.createDate에 @CreateDate 사용
- Answer에 Question 연관관계 설정시 기존 Question과 Answer의 객체 연관관계 제거
- Answer.delete()를 private method로 변경
- CannotDeleteException이 RuntimeException을 상속하여 명시적으로 처리하지 않도록 수정

haneepark committed:
refactor: 코드 리뷰 의견 반영

- DeleteHistory의 ofQuestion(), ofAnswer()를 private method로 변경
- 필요없는 코멘트 삭제

haneepark committed:
refactor: Cascade, Fetch 전략 추가 및 테스트 격리를 위한 truncate.sql 추가

haneepark commented on src/test/java/qna/domain/QuestionTest.java:
```diff
@@ -31,6 +41,18 @@ void save() {
+        final User user = users.save(UserTest.JAVAJIGI);
+        final Question question = Q1.writeBy(user);
+        question.addAnswer(AnswerTest.A1);
+        final Question savedQuestion = questions.save(question);
```
@minsungryu 
이 라인에서, 이런 에러 메시지와 함께 실패하고있는데요,

```
Caused by: javax.persistence.EntityNotFoundException: Unable to find qna.domain.User with id 1
```

`UserTest.JAVAJIGI` 객체를 save 했는데 왜 찾지 못하는걸까요?
truncate.sql로 테스트 격리도 해주었으니 id가 꼬인 것도 아닌데. 고민을 해봤지만 원인을 잘 모르겠습니다. 도움 주시면 감사하겠습니다!

minsungryu replied:
`AnswerTest.A1`이 참조하는 `User` 개체가 달라서 그런것으로 보이네요~
`UserTest.JAVAJIGI`의 ID가 1로 고정되어 있는데요,
테스트들이 실행되면서 DB에 저장하는 코드들이 있는데 테스트가 롤백 되더라도 PK값은 계속 증가하게 됩니다.
그래서 46라인의 `user`와 48라인의 `user`가 각각 다른 ID를 갖게 되어 다르다고 인식됩니다.
연관 관계 매핑을 했으니 JPA 입장에서는 ID가 1인 `user`를 조회해와서 `Answer` 개체에 매핑을 시도하게 될텐데요
이 때 DB에는 ID가 1인 `user`가 없다보니 에러가 나는 것입니다.

`AnswerTest.A1`의 `user`를 변경해주면 되긴 하지만 그렇게되면 다른 테스트가 영향을 받아 실패하게 되는데요
테스트 픽스처에 대해서는 아래 포스트를 읽어보시고 각 테스트 독립적으로 실행될 수 있도록 구성해보시는 것을 추천드립니다.
- https://jojoldu.tistory.com/611?category=1011740
- https://jobjava00.github.io/book/xunit/08/
- https://jobjava00.github.io/book/xunit/09/

끝으로 이 테스트의 마지막 단언문은 실패할텐데 `Answer` 엔티티에 `equals & hashcode` 구현을 해주시면 될 것 같습니다.

haneepark replied:
이해했습니다. 감사합니다 :)

haneepark commented:
의견 주신부분 반영 + 테스트 질문 하나 남겼습니다. 확인 부탁드립니다!

minsungryu commented on src/main/java/qna/domain/Answer.java:
```diff
@@ -55,6 +55,9 @@ public boolean isOwner(User writer) {
 
     public void toQuestion(Question question) {
+        if (this.question != null) {
+            question.getAnswers().remove(this);
```
`removeAnswer` 메서드를 만들어보는 것은 어떨까요?

minsungryu reviewed (approved):
하늬님, ATDD 미션 완료하느라 고생 많으셨습니다 👍 
클린 코드를 구현하는 역량이 뛰어나신 것 같습니다 💯 
다음 미션도 잘 진행하시길 바래요~ 화이팅입니다.
